internal/bridge/client_test.go

```diff
@@ -536,3 +542,957 @@ func TestClientCreateInstallation(t *testing.T) {
+			w.WriteHeader(http.StatusOK)
+			_, _ = w.Write(responsePayloadJson)
+		}))
+		t.Cleanup(server.Close)
+
+		client := NewClient(server.URL, "")
+		assert.Equal(t, client.BaseURL.String(), server.URL)
+
+		clusterUpdate, err := client.UpdateCluster(context.Background(), apiKey, clusterId, clusterUpdateRequestPayload)
+		assert.NilError(t, err)
+		assert.Equal(t, clusterUpdate.ClusterName, clusterApiResource.ClusterName)
+		// TODO: check every field?
+	})
+
+	t.Run("ErrorResponse", func(t *testing.T) {
+		responsePayloadJson, err := json.Marshal(clusterApiResource)
+		assert.NilError(t, err)
+
+		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.WriteHeader(http.StatusBadRequest)
+			_, _ = w.Write(responsePayloadJson)
+		}))
+		t.Cleanup(server.Close)
+
+		client := NewClient(server.URL, "")
+		assert.Equal(t, client.BaseURL.String(), server.URL)
+
+		_, err = client.UpdateCluster(context.Background(), apiKey, clusterId, clusterUpdateRequestPayload)
+		assert.Check(t, err != nil)
+		assert.ErrorContains(t, err, "400 Bad Request")
+	})
+}
+
+func TestGetClusterRole(t *testing.T) {
+	apiKey := "1234"
+	clusterId := "5678"
+	teamId := "9012"
+	roleName := "application"
+	clusterRoleApiResource := &ClusterRoleApiResource{
+		AccountEmail: "test@email.com",
+		AccountId:    "12345678",
+		ClusterId:    clusterId,
+		Flavor:       "chocolate",
+		Name:         roleName,
+		Password:     "password",
+		Team:         teamId,
+		URI:          "connection-string",
+	}
+
+	t.Run("WeSendCorrectData", func(t *testing.T) {
+		responsePayloadJson, err := json.Marshal(clusterRoleApiResource)
+		assert.NilError(t, err)
+
+		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			assert.Equal(t, r.Method, "GET", "Expected GET method")
+			assert.Equal(t, r.URL.Path, "/clusters/"+clusterId+"/roles/"+roleName,
+				fmt.Sprintf("Expected path to be '/clusters/%s/roles/%s'", clusterId, roleName))
+			assert.Equal(t, r.Header.Get("Authorization"), "Bearer "+apiKey, "Expected Authorization header to contain api key.")
+
+			w.WriteHeader(http.StatusOK)
+			_, _ = w.Write(responsePayloadJson)
+		}))
+		t.Cleanup(server.Close)
+
+		client := NewClient(server.URL, "")
+		assert.Equal(t, client.BaseURL.String(), server.URL)
+
+		_, err = client.GetClusterRole(context.Background(), apiKey, clusterId, roleName)
+		assert.NilError(t, err)
+	})
+
+	t.Run("OkResponse", func(t *testing.T) {
+		responsePayloadJson, err := json.Marshal(clusterRoleApiResource)
+		assert.NilError(t, err)
+
+		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.WriteHeader(http.StatusOK)
+			_, _ = w.Write(responsePayloadJson)
+		}))
+		t.Cleanup(server.Close)
+
+		client := NewClient(server.URL, "")
+		assert.Equal(t, client.BaseURL.String(), server.URL)
+
+		clusterRole, err := client.GetClusterRole(context.Background(), apiKey, clusterId, roleName)
```

Comment:
The test is written so that the http server sends a hard-coded response... In these tests, we aren't actually mocking the API, we are simply ensuring that we are sending the right data to Bridge (in these tests, a fake http server) and that when we get a 200/400 response, that we return the right things.

Reply:
hm - so `NewClient` here is using the mock functions that you just moved to the new file?

Reply:
No, `NewClient` here returns a `bridge.Client` but it sets the `BaseURL` to the test http server's (defined above on L1361) URL so that requests get sent to the test http server rather than to the Bridge API... The mock bridge api is used by the tests in `crunchybridgecluster_controller_test.go`. They replace the `NewClient` function with one that returns a `TestBridgeClient` rather than a `bridge.Client`.